flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java

```diff
@@ -0,0 +1,328 @@
+
+import org.apache.flink.api.common.JobStatus;
+import org.apache.flink.runtime.checkpoint.CheckpointCoordinator;
+import org.apache.flink.runtime.checkpoint.CheckpointScheduling;
+import org.apache.flink.runtime.concurrent.FutureUtils;
+import org.apache.flink.runtime.execution.ExecutionState;
+import org.apache.flink.runtime.executiongraph.ArchivedExecutionGraph;
+import org.apache.flink.runtime.executiongraph.ExecutionGraph;
+import org.apache.flink.runtime.executiongraph.TaskExecutionStateTransition;
+import org.apache.flink.runtime.scheduler.ExecutionGraphHandler;
+import org.apache.flink.runtime.scheduler.OperatorCoordinatorHandler;
+import org.apache.flink.util.FlinkException;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+
+import javax.annotation.Nullable;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Tracks a "stop with savepoint" operation. The incoming "savepointFuture" is coming from the
+ * {@link CheckpointCoordinator}, which takes care of triggering a savepoint, and then shutting down
+ * the job (on success).
+ *
+ * <p>This state is tracking the future to act accordingly on it. The savepoint path (= the result
+ * of the operation) is made available via the "operationFuture" to the user. This operation is only
+ * considered successfully if the "savepointFuture" completed successfully, and the job reached the
+ * terminal state FINISHED.
+ */
+class StopWithSavepoint extends StateWithExecutionGraph {
+
+    private final Context context;
+    private final ClassLoader userCodeClassLoader;
+
+    private final CompletableFuture<String> operationFuture;
+
+    private final CheckpointScheduling checkpointScheduling;
+
+    private boolean hasFullyFinished = false;
+
+    @Nullable private String savepoint = null;
+
+    @Nullable private Throwable operationFailureCause;
+
+    StopWithSavepoint(
+            Context context,
+            ExecutionGraph executionGraph,
+            ExecutionGraphHandler executionGraphHandler,
+            OperatorCoordinatorHandler operatorCoordinatorHandler,
+            CheckpointScheduling checkpointScheduling,
+            Logger logger,
+            ClassLoader userCodeClassLoader,
+            CompletableFuture<String> savepointFuture) {
+        super(context, executionGraph, executionGraphHandler, operatorCoordinatorHandler, logger);
+        this.context = context;
+        this.userCodeClassLoader = userCodeClassLoader;
+        this.checkpointScheduling = checkpointScheduling;
+        this.operationFuture = new CompletableFuture<>();
+
+        FutureUtils.assertNoException(
+                savepointFuture.handle(
+                        (savepointLocation, throwable) -> {
+                            context.runIfState(
+                                    this,
+                                    () -> handleSavepointCompletion(savepointLocation, throwable),
+                                    Duration.ZERO);
```

Comment:
Maybe add a comment stating that we need this in order to not execute the callback within the constructor.